examples/multigpu_advtrain/evaluator.py

```diff
@@ -0,0 +1,333 @@
+                 nb_iter=None, dataset=None, ngpu=1):
+    model = MLP_probs(model)
+    if dataset == 'mnist':
+        nb_iter0 = 40
+        eps = .3
+    else:
+        nb_iter0 = 20
+        eps = 8./255
+    if attack_type == 'FGSM':
+        fgsm_ = {'eps': eps, 'clip_min': 0., 'clip_max': 1.}
+        fgsm = FastGradientMethod(model, sess=sess)
+        adv_x = fgsm.generate(x, **fgsm_)
+    elif attack_type == 'FGSM_randeps':
+        fgsm_ = {
+            'eps': tf.abs(tf.truncated_normal((batch_size, 1, 1, 1), 0.,
+                                              .4 / 2)),
+            'clip_min': 0., 'clip_max': 1.}
+        fgsm = FastGradientMethod(model, sess=sess)
+        adv_x = fgsm.generate(x, **fgsm_)
+    elif 'FGSM_' in attack_type:
+        eps = float(attack_type.split('_')[1]) / 255.
+        fgsm_ = {'eps': eps, 'clip_min': 0., 'clip_max': 1.}
+        fgsm = FastGradientMethod(model, sess=sess)
+        adv_x = fgsm.generate(x, **fgsm_)
+    elif attack_type == 'BasicIter':
+        attack_ = {'eps': eps, 'eps_iter': eps, 'nb_iter': 10,
+                   'clip_min': 0., 'clip_max': 1.}
+        attack = BasicIterativeMethod(model, sess=sess)
+        adv_x = attack.generate(x, **attack_)
+    elif 'BasicIter_' in attack_type:
+        eps = float(attack_type.split('_')[1]) / 255.
+        eps_i = 0.1
+        nb_iter = min(30, max(1, int(eps / eps_i * 10)))
+        attack_ = {'eps': eps, 'eps_iter': eps_i, 'nb_iter': nb_iter,
+                   'clip_min': 0., 'clip_max': 1.}
+        logging.info('%s: %.4f %.4f %d' % (attack_type, eps, eps_i, nb_iter))
+        attack = BasicIterativeMethod(model, sess=sess)
+        adv_x = attack.generate(x, **attack_)
+    elif attack_type == 'PGDcl':
+        # from paper https://arxiv.org/pdf/1706.06083.pdf
+        if nb_iter is None:
+            nb_iter = 40
+        print('nb_iter: %d, eps: %.2f' % (nb_iter, eps))
+        attack_ = {'eps': eps, 'eps_iter': 0.01, 'nb_iter': nb_iter,
+                   'clip_min': 0., 'clip_max': 1., 'pgd': False}
+        attack = ProjectedGradientDescentMethod(model, sess=sess)
+        adv_x = attack.generate(x, **attack_)
+    elif 'PGDpgd' in attack_type:
+        # from paper https://arxiv.org/pdf/1706.06083.pdf
+        if nb_iter is None:
+            nb_iter = nb_iter0
+        print('nb_iter: %d, eps: %.2f' % (nb_iter, eps))
+        attack_ = {'eps': eps, 'eps_iter': 0.01, 'nb_iter': nb_iter,
+                   'clip_min': 0., 'clip_max': 1., 'pgd': True}
+        if attack_type == 'PGDpgd_y':
+            attack_['y'] = y
+        if 'multigpu' in attack_type:
+            attack_['multigpu'] = True
+            attack_['ngpu'] = ngpu
+        attack = ProjectedGradientDescentMethod(model, sess=sess)
+        adv_x = attack.generate(x, **attack_)
+    else:
+        raise Exception('Attack %s not defined.' % attack_type)
+    return adv_x
+
+
+class Evaluator(object):
```

Comment:
Add a class-level docstring saying what the class is for / documenting the parameters of the init method

Reply:
Fixed in 5f7bc3e and 44356a0.